app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,555 @@
+        LocationSettingsRequest.Builder();
+    builder.addLocationRequest(locationRequest);
+
+    task = LocationServices.getSettingsClient(this).checkLocationSettings(builder.build());
+
+    locationSettingsResponseBuilder();
+  }
+
+  private void locationSettingsResponseBuilder() {
+    task.addOnCompleteListener(task -> {
+      try {
+        LocationSettingsResponse response = task.getResult(ApiException.class);
+        // All location settings are satisfied. The client can initialize location
+        // requests here.
+
+        createHotspotIntent(ACTION_TURN_ON_AFTER_O);
+
+        //}
+      } catch (ApiException exception) {
+        switch (exception.getStatusCode()) {
+          case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:
+            // Location settings are not satisfied. But could be fixed by showing the
+            // user a dialog.
+            try {
+              // Cast to a resolvable exception.
+              ResolvableApiException resolvable = (ResolvableApiException) exception;
+              // Show the dialog by calling startResolutionForResult(),
+              // and check the result in onActivityResult().
+              resolvable.startResolutionForResult(
+                  ZimHostActivity.this,
+                  LOCATION_SETTINGS_PERMISSION_RESULT);
+            } catch (IntentSender.SendIntentException e) {
+              // Ignore the error.
+            } catch (ClassCastException e) {
+              // Ignore, should be an impossible error.
+            }
+            break;
+          case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
+            // Location settings are not satisfied. However, we have no way to fix the
+            // settings so we won't show the dialog.
+            break;
+          default:
+            break;
+        }
+      }
+    });
+  }
+
+  //Advice user to turn on hotspot manually for API<26
+  private void startHotspotManuallyDialog() {
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+    builder.setPositiveButton(getString(R.string.go_to_settings_label),
+        (dialog, id) -> launchTetheringSettingsScreen());
+
+    builder.setNeutralButton(getString(R.string.hotspot_dialog_neutral_button), (dialog, id) -> {
+
+      progressDialog =
+          ProgressDialog.show(this, getString(R.string.progress_dialog_starting_server), "",
+              true);
+      progressDialog.show();
+
+      startFlowable();
+    });
+
+    builder.setTitle(getString(R.string.hotspot_dialog_title));
+    builder.setMessage(
+        getString(R.string.hotspot_dialog_message)
+    );
+    AlertDialog dialog = builder.create();
+    dialog.show();
+  }
+
+  //Keeps checking if hotspot has been turned using the ip address with an interval of 1 sec
+  //If no ip is found after 15 seconds, dismisses the progress dialog
+  private void startFlowable() {
+    Flowable.fromCallable(() -> ServerUtils.getIp())
```

Comment:
did the `flowable.timer` solution not work out? The hitting an exception repeatedly thing is a little less beautiful in my mind.
I'd also do a `take(1)` after a `filter(ipAddress -> ipAddress.isNotEmpty())`as opposed to converting to a Single

Reply:
We don't know after what duration we'll get a valid ip so we have not used timer as it will not hit the method again. If we use timer we have to give a large delay which is not user friendly.

Reply:
Yeah I am a big dumb dumb.
*`Flowable.interval(1, TimeUnit.Seconds)`

Reply:
I used interval but mapping time to method call was not looking clean that's why I changed it to callable.

Reply:
```
Flowable.interval(1, TimeUnit.SECONDS)
      .timeout(15, TimeUnit.SECONDS)
      .map( __-> ServerUtils.getIp())
      .filter( s -> !s.isEmpty())
      .take(1)
      .subscribe(
        s -> {//success stuff},
        e -> {//failure stuff}
      );
```
I think it looks neater and stopping ServerUtils throwing that exception seem like a win to me. Also if server utils is going to be throwing exceptions it should be a part of its signature.
There is no way of knowing from the original code why you have to retry every 1 seconds from a failure

Reply:
Also all of this seem wrong when there are methods like this
```
  @Override public void onServerStarted(@NonNull String ipAddress) {
    this.ip = ipAddress;
    layoutServerStarted();
  }
```
The hotspotService or one of its dependencies should be doing this and reporting events back, the client is doing an awful lot of work when all it should be doing is showing what the hotspot service tells it to render

Reply:
@iadeelzafar Work on these too.

Reply:
@macgills I've now
Fixed flowable
Also shifted all the flowable code to `WebServerHelper`
Check this 6c75fefdb7afc070179ecc4639a50cc603e3723f